mathics/core/expression.py

```diff
@@ -1014,18 +1014,20 @@ def flatten_callback(new_elements, old):
 
         new._timestamp_cache(evaluation)
 
-        # Step 5: Would the operation benefit running in separate threads?
+        # Step 5: Must we need to thread-rewrite the expression?
         #
-        # We allow threading in when head has the ``Listable``
-        # Attribute.  Here ``Expression.thread`` changes:
-        #  ``F[{a,b,c,...}]`` to:
+        # Threading is needed when head has the ``Listable``
+        # Attribute.  ``Expression.thread`` rewrites the expression:
+        #  ``F[{a,b,c,...}]`` as:
         #  ``{F[a], F[b], F[c], ...}``.
 
-        # TODO: For a small number of arguments threading is just overhead.
-        # For zero or one argument this is always the case.
-        # For two arguments, note that many listable functions are binary operators
-        # so two arguments is really the same as one argument and threading is overhead.
+        # Note: Threading here is different from Python or OS threads,
+        # even though the intent of this attribute was to allow for
```

Comment:
I would say that ``Thread`` in WL does not have anything to do with OS threads. In order to follow the semantics of the language, {F[a],F[b],...} -in general- must be evaluated sequentially.

Reply:
This sentence from https://reference.wolfram.com/language/ref/Listable.html leads me to believe that parallelization is the motivation: 

> Arguments that are not lists are copied as many times as there are elements in the lists.

In particular, this means that the rewrite process can be done in parallel, and if the list is _very_ large there may be a benefit here. 

Also, from the Mathematica V5 book section 3.8.4 Operations on Scalars, Vectors and Matrices:

> Most mathematical functions in Mathematica are set up to apply themselves separately to each element in a list. This is true in particular of all functions that carry the attribute Listable. 

Another Language that Wolfram says he took inspiration from is APL, and this is definitely a design feature of that language.

Reply:
Interesting... that maybe works for atomic leaves. However, consider this:
```
In[6]:= Sin[{a=1,a=a+1,a=a+1}]                                                  

Out[6]= {Sin[1], Sin[2], Sin[3]}
```

Reply:
The list copy is done in parallel. The _evaluation_ or _eval()_ part isn't.

Reply:
See: https://reference.wolfram.com/language/ref/Parallelize.html
`Scope -> Listable functions`